File: LightReading/Tool/LRStarViewTool/LRStarView的OC版本/LRStarView.m
//
//  LRStarView.m
//  LightReading
//
//  Created by zhifu360 on 2018/7/27.
//  Copyright © 2018年 智富金融. All rights reserved.
//

#import "StarView.h"

#define LRStarViewRGB(r,g,b) [UIColor colorWithRed:r/255.0f green:g/255.0f blue:b/255.0f alpha:1.0f]
#define LRStarViewRGBA(r,g,b,a) [UIColor colorWithRed:r/255.0f green:g/255.0f blue:b/255.0f alpha:a]

@implementation StarView

///LRStarView的OC版本
- (instancetype)initWithFrame:(CGRect)frame {
    
    self = [super initWithFrame:frame];
    if (self) {
        self.backgroundColor = [UIColor clearColor];
        self.starSize = 13.0f;
        self.emptyColor = LRStarViewRGB(167, 167, 167);
        self.fullColor = LRStarViewRGB(255, 121, 22);
        self.maxStar = 100;
    }
    return self;
}

- (void)drawRect:(CGRect)rect {
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    NSString *stars = @"★★★★★";
    
    rect = self.bounds;
    UIFont *font = [UIFont boldSystemFontOfSize:self.starSize];
    NSDictionary *attributes = @{NSFontAttributeName:font};
    CGSize starSize = [stars sizeWithAttributes:attributes];
    rect.size = starSize;
    [self.emptyColor set];
    [stars drawInRect:rect withAttributes:attributes];
    
    CGRect clip = rect;
    clip.size.width = clip.size.width * self.showStar / self.maxStar;
    CGContextClipToRect(context, clip);
    [self.fullColor set];
    [stars drawInRect:rect withAttributes:attributes];
}

@end

File: LightReading/Public/LRStarView.swift
//
//  LRStarView.swift
//  LightReading
//
//  Created by zhifu360 on 2018/7/26.
//  Copyright © 2018年 智富金融. All rights reserved.
//

import UIKit

class LRStarView: UIView {

    var starSize:CGFloat = 0.0//根据字体大小来确定星星的大小
    
    var maxStar:CGFloat = 0.0//总共的长度
    
    var showStar:CGFloat = 0.0//需要显示的星星的长度
    
    var emptyColor:UIColor!//未点亮时候的颜色
    
    var fullColor:UIColor!//点亮的星星的颜色
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.createView()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func createView() {
        //初始化
        self.backgroundColor = .clear
        self.starSize = 13.0
        self.emptyColor = RGB(167, 167, 167)
        self.fullColor = RGB(255, 121, 22)
        self.maxStar = 5.0
    }
    
    override func draw(_ rect: CGRect) {
        let context = UIGraphicsGetCurrentContext()
        
        let stars = "★★★★★"
        
        var tmpRect = rect
        tmpRect = self.bounds
        let starSize = (stars as NSString).size(withAttributes: [NSAttributedStringKey.font:DRBoldFont(self.starSize)])
        tmpRect.size = starSize
        print("tmpRect = \(tmpRect)")
        self.emptyColor.set()
        (stars as NSString).draw(in: tmpRect, withAttributes: [NSAttributedStringKey.font:DRBoldFont(self.starSize)])
        
        var clip:CGRect = rect
        clip.size.width = clip.size.width * self.showStar / self.maxStar
        context?.clip(to: clip)
        self.fullColor.set()
        (stars as NSString).draw(in: tmpRect, withAttributes: [NSAttributedStringKey.font:DRBoldFont(self.starSize)])
        self.backgroundColor = .clear
    }
}










